datahub/company/tasks.py

```diff
@@ -0,0 +1,79 @@
+from celery import shared_task
+from celery.utils.log import get_task_logger
+from dateutil.relativedelta import relativedelta
+from django.db.models import OuterRef, Q, Subquery
+from django.utils import timezone
+from django_pglocks import advisory_lock
+
+from datahub.company.constants import AUTOMATIC_COMPANY_ARCHIVE_FEATURE_FLAG
+from datahub.company.models.company import Company
+from datahub.feature_flag.utils import is_feature_flag_active
+from datahub.interaction.models import Interaction
+
+logger = get_task_logger(__name__)
+
+
+def _automatic_company_archive(task, batch_size, simulate):
+
+    _8y_ago = timezone.now() - relativedelta(years=8)
+    _3m_ago = timezone.now() - relativedelta(months=3)
+
+    latest_interaction = Interaction.objects.filter(
+        company=OuterRef('pk'),
+    ).order_by('-date')
+
+    companies_to_be_archived = Company.objects.annotate(
+        latest_interaction_date=Subquery(
+            latest_interaction.values('date')[:1],
+        ),
+    ).filter(
+        Q(latest_interaction_date__date__lt=_8y_ago) | Q(latest_interaction_date__isnull=True),
+        archived=False,
+        duns_number__isnull=True,
+        orders__isnull=True,
+        created_on__lt=_3m_ago,
+        modified_on__lt=_3m_ago,
+    ).all()[:batch_size]
```

Comment:
Minor point – `.all()` shouldn't be needed here:

```suggestion
    )[:batch_size]
```